osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,229 @@
 use crate::osm::NodeID;
+use crate::types::Movement;
+use crate::Direction;
 use crate::IntersectionComplexity::*;
-use crate::{Direction, IntersectionComplexity, StreetNetwork};
+use crate::{
+    ConflictType, DrivingSide, IntersectionComplexity, RestrictionType, Road, StreetNetwork,
+};
+use std::cmp::{max, min};
+use std::collections::BTreeMap;
 
 /// Determines the initial complexity of all intersections. Intersections marked "Crossing" are
 /// considered "unclassified" and will be updated with a guess, others will be left unchanged.
 pub fn classify_intersections(streets: &mut StreetNetwork) {
-    let mut changes: Vec<(NodeID, IntersectionComplexity)> = Vec::new();
+    let mut changes: Vec<_> = Vec::new();
     for (id, inter) in &streets.intersections {
         if let Crossing = inter.complexity {
             changes.push((*id, guess_complexity(streets, id)));
         }
     }
 
-    for (id, complexity) in changes.into_iter() {
-        streets.intersections.get_mut(&id).unwrap().complexity = complexity;
+    for (id, (complexity, conflict_level, movements)) in changes {
+        let intersection = streets.intersections.get_mut(&id).unwrap();
+        intersection.complexity = complexity;
+        intersection.conflict_level = conflict_level;
+        intersection.movements = movements;
     }
 }
 
 /// Guesses the complexity of the intersection based on the connecting roads and their lanes.
 ///
 /// The existing complexity field is ignored, so be careful how you use the guessed value.
-fn guess_complexity(streets: &StreetNetwork, intersection_id: &NodeID) -> IntersectionComplexity {
-    let roads = streets.roads_per_intersection(*intersection_id);
+pub fn guess_complexity(
+    streets: &StreetNetwork,
+    intersection_id: &NodeID,
+) -> (IntersectionComplexity, ConflictType, Vec<Movement>) {
+    use ConflictType::*;
+    let roads: Vec<_> = streets
+        .roads_per_intersection(*intersection_id)
+        .iter()
+        .map(|id| &streets.roads[id])
+        .filter(|road| road.is_driveable())
+        .collect();
 
     // A terminus is characterised by a single connected road.
     if roads.len() == 1 {
-        return Terminus;
+        return (Terminus, Uncontested, Vec::new());
```

Comment:
I don't remember if we've talked about this before or not, but `Terminus Cross` is impossible, as are many others. Maybe the `IntersectionComplexity` enum cases should only embed `ConflictType` when it's possible? No need to do that in this PR, just an idea for a possible cleanup

Reply:
`IntersectionComplexity` might become `IntersectionType { MapEdge, Terminus, Connection, Fork, Intersection(ConflictType) }` or something. I've added a comment talking about the awkwardness to `IntersectionComplexity`.